zkevm-circuits/src/state_circuit/state.rs

```diff
@@ -98,453 +75,327 @@ pub struct Config<
+        // TODO: rewrite using range check gates rather than lookup
         meta.lookup_any("rw counter monotonicity", |meta| {
+            let s_enable = meta.query_fixed(s_enable, Rotation::cur());
             let rw_counter_table = meta.query_fixed(rw_counter_table, Rotation::cur());
             let rw_counter_prev = meta.query_advice(rw_counter, Rotation::prev());
             let rw_counter = meta.query_advice(rw_counter, Rotation::cur());
-            let padding = meta.query_advice(padding, Rotation::cur());
-            let is_not_padding = one.clone() - padding;
-            let q_not_first = q_memory_not_first(meta) + q_stack_not_first(meta);
 
             vec![(
-                q_not_first
-                    * is_not_padding
-                    * address_diff_is_zero.clone().is_zero_expression
+                s_enable * q_all_keys_same(meta)
                     * (rw_counter - rw_counter_prev - one.clone()), /*
                                                                      * - 1 because it needs to
                                                                      *   be strictly monotone */
                 rw_counter_table,
             )]
         });
 
-        // Memory address is in the allowed range.
+        ///////////////////////// Memory related constraints /////////////////////////
+
+        meta.create_gate("Memory operation", |meta| {
+            let mut cb = new_cb();
+            let s_enable = meta.query_fixed(s_enable, Rotation::cur());
+            let value_cur = meta.query_advice(value, Rotation::cur());
+            let is_write = meta.query_advice(is_write, Rotation::cur());
+            let q_read = one.clone() - is_write;
+
+            // 0. Unused keys are 0
+            let key2 = meta.query_advice(keys[2], Rotation::cur());
+            let key4 = meta.query_advice(keys[4], Rotation::cur());
+            cb.require_zero("key2 is 0", key2);
+            cb.require_zero("key4 is 0", key4);
+
+            // 1. First access for a set of all keys
+            //
+            // When the set of all keys changes (first access of an address in a call)
+            // - If READ, value must be 0
+            cb.require_zero(
+                "if address changes, read value should be 0",
+                q_not_all_keys_same(meta) * q_read * value_cur,
+            );
+
+            cb.gate(s_enable * q_memory(meta))
+        });
+
+        // 2. mem_addr in range
+        // TODO: rewrite this using range check gates instead of lookup
         meta.lookup_any("Memory address in allowed range", |meta| {
-            let q_memory = q_memory_first(meta) + q_memory_not_first(meta);
+            let q_memory = q_memory(meta);
             let address_cur = meta.query_advice(address, Rotation::cur());
             let memory_address_table_zero =
                 meta.query_fixed(memory_address_table_zero, Rotation::cur());
 
             vec![(q_memory * address_cur, memory_address_table_zero)]
```

Comment:
since for mem addr 0 is a valid value, so memory_address_table_zero will contain 0. So even for s_enable == false rows, the  `q_memory * address_cur` will be 0(default value) x 0(default value), it will be allowed in the lookup table.  I will add a comment on this.